File: src/main/services/mod-manager/preview.test.ts
import assert from "node:assert/strict";
import os from "node:os";
import path from "node:path";

import fse from "fs-extra";
import { afterEach, describe, it } from "vitest";

import { findPreview } from "./preview.ts";

const tempRoots: string[] = [];

afterEach(async () => {
    await Promise.all(tempRoots.splice(0).map((root) => fse.remove(root)));
});

async function makeRoot() {
    const root = await fse.mkdtemp(path.join(os.tmpdir(), "nhd-preview-"));
    tempRoots.push(root);
    return root;
}

async function writeFile(root: string, relativePath: string) {
    const filePath = path.join(root, relativePath);
    await fse.ensureDir(path.dirname(filePath));
    await fse.writeFile(filePath, "preview");
    return filePath;
}

describe("findPreview", () => {
    it("accepts a media file that is not named preview", async () => {
        const root = await makeRoot();
        const screenshot = await writeFile(root, "screenshot.png");

        assert.equal(await findPreview(root, false), screenshot);
    });

    it("prefers a preview-named file over a generic root image", async () => {
        const root = await makeRoot();
        const preview = await writeFile(root, "preview.webp");
        await writeFile(root, "cover.jpg");

        assert.equal(await findPreview(root, false), preview);
    });

    it("accepts a file whose name contains preview", async () => {
        const root = await makeRoot();
        const preview = await writeFile(root, "mod_preview.png");

        assert.equal(await findPreview(root, false), preview);
    });

    it("searches nested folders only when asked", async () => {
        const root = await makeRoot();
        const nested = await writeFile(root, "images/cover.jpg");

        assert.equal(await findPreview(root, true), nested);
        assert.equal(await findPreview(root, false), undefined);
    });

    it("prefers a root image over a nested preview in a disabled folder", async () => {
        const root = await makeRoot();
        const screenshot = await writeFile(root, "screenshot.png");
        await writeFile(root, "DISABLED Nested/preview.png");

        assert.equal(await findPreview(root, true), screenshot);
    });

    it("prefers a group-root preview over a child folder", async () => {
        const root = await makeRoot();
        const rootPreview = await writeFile(root, "preview.png");
        await writeFile(root, "Enabled Mod/preview.png");

        assert.equal(await findPreview(root, true), rootPreview);
    });

    it("prefers an enabled child folder before a disabled one", async () => {
        const root = await makeRoot();
        const enabledPreview = await writeFile(root, "Enabled Mod/nested/deeper/preview.png");
        await writeFile(root, "DISABLED Other Mod/preview.png");

        assert.equal(await findPreview(root, true), enabledPreview);
    });

    it("falls back to a disabled child folder", async () => {
        const root = await makeRoot();
        const disabledPreview = await writeFile(root, "DISABLED Other Mod/preview.png");

        assert.equal(await findPreview(root, true), disabledPreview);
    });

    it("accepts a non-preview filename in a child folder", async () => {
        const root = await makeRoot();
        const cover = await writeFile(root, "Enabled Mod/cover.jpg");

        assert.equal(await findPreview(root, true), cover);
    });

    it("recognizes a disabled preview file suffix", async () => {
        const root = await makeRoot();
        const preview = await writeFile(root, "preview.png.disabled");

        assert.equal(await findPreview(root, false), preview);
    });

    it("ignores texture-like names", async () => {
        const root = await makeRoot();
        await writeFile(root, "normal.png");
        await writeFile(root, "light.jpg");
        await writeFile(root, "material.webp");
        await writeFile(root, "diffuse.png");

        assert.equal(await findPreview(root, false), undefined);
    });
});

File: internal/mod/preview_test.go
package mod

import (
	"os"
	"path/filepath"
	"testing"
)

func writePreviewFile(t *testing.T, root, relative string) string {
	t.Helper()
	path := filepath.Join(root, filepath.FromSlash(relative))
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(path, []byte("preview"), 0o644); err != nil {
		t.Fatal(err)
	}
	return path
}

func previewEqual(t *testing.T, got *string, want string) {
	t.Helper()
	if want == "" {
		if got != nil {
			t.Fatalf("preview = %q, want nil", *got)
		}
		return
	}
	if got == nil || *got != want {
		gotValue := "<nil>"
		if got != nil {
			gotValue = *got
		}
		t.Fatalf("preview = %q, want %q", gotValue, want)
	}
}

func TestFindPreview(t *testing.T) {
	t.Parallel()

	t.Run("accepts a media file that is not named preview", func(t *testing.T) {
		t.Parallel()
		root := t.TempDir()
		screenshot := writePreviewFile(t, root, "screenshot.png")
		previewEqual(t, findPreview(root, false), screenshot)
	})

	t.Run("accepts ogg in the NTE preview finder", func(t *testing.T) {
		t.Parallel()
		root := t.TempDir()
		preview := writePreviewFile(t, root, "preview.ogg")
		previewEqual(t, findPreview(root, false), preview)
	})

	t.Run("prefers a preview-named file over a generic root image", func(t *testing.T) {
		t.Parallel()
		root := t.TempDir()
		preview := writePreviewFile(t, root, "preview.webp")
		writePreviewFile(t, root, "cover.jpg")
		previewEqual(t, findPreview(root, false), preview)
	})

	t.Run("accepts a file whose name contains preview", func(t *testing.T) {
		t.Parallel()
		root := t.TempDir()
		preview := writePreviewFile(t, root, "mod_preview.png")
		previewEqual(t, findPreview(root, false), preview)
	})

	t.Run("searches nested folders only when asked", func(t *testing.T) {
		t.Parallel()
		root := t.TempDir()
		nested := writePreviewFile(t, root, "images/cover.jpg")
		previewEqual(t, findPreview(root, true), nested)
		previewEqual(t, findPreview(root, false), "")
	})

	t.Run("prefers a root image over a nested preview in a disabled folder", func(t *testing.T) {
		t.Parallel()
		root := t.TempDir()
		screenshot := writePreviewFile(t, root, "screenshot.png")
		writePreviewFile(t, root, "DISABLED Nested/preview.png")
		previewEqual(t, findPreview(root, true), screenshot)
	})

	t.Run("prefers a group-root preview over a child folder", func(t *testing.T) {
		t.Parallel()
		root := t.TempDir()
		rootPreview := writePreviewFile(t, root, "preview.png")
		writePreviewFile(t, root, "Enabled Mod/preview.png")
		previewEqual(t, findPreview(root, true), rootPreview)
	})

	t.Run("prefers an enabled child folder before a disabled one", func(t *testing.T) {
		t.Parallel()
		root := t.TempDir()
		enabledPreview := writePreviewFile(t, root, "Enabled Mod/nested/deeper/preview.png")
		writePreviewFile(t, root, "DISABLED Other Mod/preview.png")
		previewEqual(t, findPreview(root, true), enabledPreview)
	})

	t.Run("falls back to a disabled child folder", func(t *testing.T) {
		t.Parallel()
		root := t.TempDir()
		disabledPreview := writePreviewFile(t, root, "DISABLED Other Mod/preview.png")
		previewEqual(t, findPreview(root, true), disabledPreview)
	})

	t.Run("does not search child folders at depth one", func(t *testing.T) {
		t.Parallel()
		root := t.TempDir()
		writePreviewFile(t, root, "Enabled Mod/preview.png")
		previewEqual(t, findGroupPreview(root, 1), "")
	})

	t.Run("caps child preview search at rust depth three", func(t *testing.T) {
		t.Parallel()
		root := t.TempDir()
		found := writePreviewFile(t, root, "Enabled Mod/nested/deeper/preview.png")
		writePreviewFile(t, root, "Enabled Mod/a/b/c/d/too-deep.png")
		previewEqual(t, findPreview(root, true), found)
	})

	t.Run("ignores previews deeper than rust child depth three", func(t *testing.T) {
		t.Parallel()
		root := t.TempDir()
		writePreviewFile(t, root, "Enabled Mod/a/b/c/d/preview.png")
		previewEqual(t, findPreview(root, true), "")
	})

	t.Run("accepts a non-preview filename in a child folder", func(t *testing.T) {
		t.Parallel()
		root := t.TempDir()
		cover := writePreviewFile(t, root, "Enabled Mod/cover.jpg")
		previewEqual(t, findPreview(root, true), cover)
	})

	t.Run("recognizes a disabled preview file suffix", func(t *testing.T) {
		t.Parallel()
		root := t.TempDir()
		preview := writePreviewFile(t, root, "preview.png.disabled")
		previewEqual(t, findPreview(root, false), preview)
	})

	t.Run("ignores texture-like names", func(t *testing.T) {
		t.Parallel()
		root := t.TempDir()
		writePreviewFile(t, root, "normal.png")
		writePreviewFile(t, root, "light.jpg")
		writePreviewFile(t, root, "material.webp")
		writePreviewFile(t, root, "diffuse.png")
		previewEqual(t, findPreview(root, false), "")
	})
}
